src/preferences/common/preference-service.ts

```diff
@@ -0,0 +1,127 @@
+    // get<T>(preferenceName: string, defaultValue?: T): Promise<T>;
+    // set<T>(preferenceName: string, value: T): void;
+    // remove<T>(preferenceName: string): T | undefined;
+
+}
+
+export interface PreferenceServiceClient {
+    onDidChangePreference(event: PreferenceChangedEvent): void
+}
+
+export class JsonPreferenceService implements PreferenceService {
+
+    // A minimal part of the interface.
+    readonly onDidChangePreference: Event<PreferenceChangedEvent>;
+    prefs: Promise<any>; // Preferences cache
+    wsRoot: Promise<string>;
+    private prefsChangedListeners: Emitter<PreferenceChangedEvent> = new Emitter();
+
+    constructor(
+        @inject(FileSystem) protected readonly fileSystem: FileSystem,
+        @inject(FileSystemWatcher) protected readonly watcher: FileSystemWatcher,
+        readonly prefFileRelativeUri: string) {
+        this.wsRoot = fileSystem.getWorkspaceRoot().then((fileStat: FileStat) => {
+            return Promise.resolve(fileStat.uri);
+        })
+
+        watcher.onFileChanges(event => {
+            this.arePreferencesAffected(event).then((arePrefsAffected) => {
+                if (arePrefsAffected) {
+                    this.reconcilePreferences();
+                }
+            })
+        });
+
+        this.reconcilePreferences();
+    }
+
+    /**
+     * Checks to see if the preference file was modified
+     */
+    arePreferencesAffected(event: FileChangesEvent): Promise<boolean> {
+        return new Promise<boolean>((resolve) => {
+            this.wsRoot.then((root) => {
+                for (let change of event.changes) {
+                    if (change.uri === this.wsRoot + "/" + this.prefFileRelativeUri) {
+                        resolve(true);
+                    }
+                }
+                resolve(false);
+            })
+        })
+    }
+
+    /**
+     * Read preferences
+     */
+    reconcilePreferences(): void {
+        this.wsRoot.then((root) => {
+            const uri = root + "/" + this.prefFileRelativeUri;
+            this.fileSystem.resolveContent(uri).then(({ stat, content }) => {
+                let newPrefs = JSON.parse(content) // Might need a custom parser because comments and whatnot?
+                if (this.prefs !== newPrefs) {
+                    // this.prefsChangedListeners.fire();
+                    // Fire events
+                }
+                this.prefs = new Promise<any>((resolve) => {
+                    resolve(newPrefs);
+                })
+            });
+        })
+    }
+
+    has(preferenceName: string): Promise<boolean> {
+        return new Promise<boolean>((resolve) => {
+            this.prefs.then((prefs) => {
+                if (prefs[preferenceName]) {
+                    resolve(true);
+                }
+                resolve(false);
+            })
+        });
+    }
+
+    get<T>(preferenceName: string): Promise<T | undefined> {
+
+        return new Promise<any>((resolve) => {
```

Comment:
While testing that with 
```
it should be protected readonly prefs: object | undefined;
 ```

I had an error of

```
Element implicitly has an 'any' type because type '{}' has no index signature
```
and the only way (I found) to fix it was to use 

```
    protected prefs: { [key: string]: any } | undefined; // Preferences cache
```

WDYT?